src/setup/rgw.py

```diff
@@ -14,77 +14,219 @@
         return 0
 
     @staticmethod
-    def prepare(config_path: str):
-        """
-        Performs prepare phase operations.
-        """
+    def prepare(conf: MappedConf):
+        """Perform prepare operations."""
+
+        Log.info('Prepare phase started.')
+
+        try:
+            hostname = conf.get(f'node>{Rgw._machine_id}>hostname')
+            host_ip = socket.gethostbyname(hostname)
+            ceph_tmpl_idx = 'ceph_conf_tmpl'
+            ceph_tmpl_url = f'ini://{CEPH_CONF_TMPL}'
+            # Load template ceph.conf file and updated required field.
+            kv_list = [
+                ('global>fsid', str(uuid.uuid1())),
+                ('global>mon host', host_ip)]
+            Rgw._update_config(ceph_tmpl_idx, ceph_tmpl_url, kv_list)
+
+            # Copy ceph_conf_tmpl_file to /etc/ceph/ dir.
+            shutil.copyfile(CEPH_CONF_TMPL, CEPH_CONF)
+            Log.info(f'{CEPH_CONF_TMPL} config copied to {CEPH_CONF}.')
+
+            # create admin keyring.
+            cmd = f"sudo ceph-authtool --create-keyring {CEPH_ADMIN_KEYRING} \
+                --gen-key -n client.admin --cap mon 'allow *' --cap osd 'allow *'"
+            _, err, rc, = SimpleProcess(cmd).run()
+            if rc != 0:
+                raise SetupError(errno.EINVAL, f'"{cmd}" failed with error {err}.')
+        except Exception as e:
+            raise SetupError(errno.EINVAL, f'Error ocurred while fetching node ip, {e}')
+
+        Log.info('Prepare phase completed.')
 
         return 0
 
     @staticmethod
-    def config(config_path: str):
-        """
-        Performs configurations.
-        """
+    def config(conf: MappedConf):
+        """Performs configurations."""
+
+        Log.info('Config phase started.')
 
+        # TODO: Inside [client.rgw.<hostname>] section and
+        # add motr,hax endpoints and rgw client related parameters.
+        Log.info('Config phase completed.')
         return 0
 
     @staticmethod
-    def init(config_path: str):
-        """
-        Perform initialization.
-        """
-
+    def init(conf: MappedConf):
+        """Perform initialization."""
+
+        Log.info('Init phase started.')
+        # TODO: Create admin user.
+        # admin user should be created only on one node.
+        # 1. While creating admin user, global lock created in consul kv store.
+        # (rgw_consul_index, cortx>rgw>volatile>rgw_lock, machine_id)
+        # 2. Before creating admin user.
+        #    a. Check for rgw_lock in consul kv store.
+        #    b. Create user only if lock value is None/machine-id.
+
+        rgw_lock = False
+        rgw_lock_key = 'component>rgw>volatile>rgw_lock'
+        rgw_consul_idx = 'rgw_consul_idx'
+        # Get consul url from cortx config.
+        consul_url = Rgw._get_consul_url(conf)
+        # Check for rgw_lock in consul kv store.
+        Log.info('Checking for rgw lock in consul kv store.')
+        Conf.load(rgw_consul_idx, consul_url)
+
+        while(True):
```

Comment:
Why we need while loop? Checks in try-catch block would be needed only once to see if lock is available, can be acquired, is already self acquired or some other machine.

Reply:
This code will execute parallelly on all nodes, and  in first iteration for all node lock key is none(if if in case try-catch block code executed same time on each node), then all node tries to update lock-key, we have added time.sleep(3) so that after updation it will wait for sometime and in next iteration all nodes will get lock value as node-id of node who has updated lock key at last. and then only that node will perform user creation operation.

This is what I understood from s3 locking code. 
@sachitanands Please correct me if I am wrong.

Reply:
Hmm ok. better add code comments here then, so we dont loose the purpose of this loop & other logic